Use new production endpoint

tameware opened the pull request:
Replaced the old GCP endpoint with the current one and added a new make target, `curl_prod`.

suokko commented:
Hi

On Thu, 9 Jul 2020, 17:44 Kiat, <notifications@github.com> wrote:

> *@kiat-huang* requested changes on this pull request.
>
> There's errors when I clone master, checkout this PR and run "make build"
>
> kiat@newton:~/hack/DDS$ git checkout master
> Switched to branch 'master'
> Your branch is up-to-date with 'origin/master'.
>
> kiat@newton:~/hack/DDS$ gh pr checkout 65
> Switched to branch 'new-prod-url'
>
> kiat@newton:~/hack/DDS$ git st
> On branch new-prod-url
> nothing to commit, working tree clean
>
> kiat@newton:~/hack/DDS$ git remote -v
> origin	git@github.com:online-bridge-hackathon/DDS.git (fetch)
> origin	git@github.com:online-bridge-hackathon/DDS.git (push)
>
> kiat@newton:~/hack/DDS$ make build
> git submodule update --init --rebase
> fatal: repository '/home/kiat/Bridge/Hackathon/online-bridge-hackathon/libdds' does not exist
> fatal: clone of '/home/kiat/Bridge/Hackathon/online-bridge-hackathon/libdds' into submodule path '/home/kiat/Bridge/Hackathon/code/DDS/libdds' failed
>
>
Cloning local repository is expected to fail. Git makes it easy to use a
local tree for all the work. But if you really need multiple trees then you
can clone the remote repository with --reference parameter.

Failed to clone 'libdds'. Retry scheduled
> fatal: repository '/home/kiat/Bridge/Hackathon/online-bridge-hackathon/libdds' does not exist
> fatal: clone of '/home/kiat/Bridge/Hackathon/online-bridge-hackathon/libdds' into submodule path '/home/kiat/Bridge/Hackathon/code/DDS/libdds' failed
> Failed to clone 'libdds' a second time, aborting
> Makefile:29: recipe for target 'libdds-update' failed
> make: *** [libdds-update] Error 1
>
> kiat@newton:~/hack/DDS$ lr
> total 68
> -rw-rw-r-- 1 kiat kiat     6 Jul  9 15:35 VERSION
> -rw-rw-r-- 1 kiat kiat   176 Jul  9 15:35 requirements.txt
> -rw-rw-r-- 1 kiat kiat 11415 Jul  9 15:35 LICENSE
> drwxrwxr-x 2 kiat kiat  4096 Jul  9 15:35 libdds
> -rw-rw-r-- 1 kiat kiat  1207 Jul  9 15:35 Dockerfile
> -rw-rw-r-- 1 kiat kiat   973 Jul  9 15:35 CONTRIBUTING.md
> -rw-rw-r-- 1 kiat kiat    25 Jul  9 15:35 _config.yml
> drwxrwxr-x 3 kiat kiat  4096 Jul  9 15:35 chart
> drwxrwxr-x 2 kiat kiat  4096 Jul  9 15:39 test
> drwxrwxr-x 2 kiat kiat  4096 Jul  9 15:39 src
> -rw-rw-r-- 1 kiat kiat  2552 Jul  9 15:39 README.md
> -rw-rw-r-- 1 kiat kiat  4561 Jul  9 15:39 README.gcp.md
> -rw-rw-r-- 1 kiat kiat  2459 Jul  9 15:39 Makefile
> drwxrwxr-x 2 kiat kiat  4096 Jul  9 15:39 frontend
> kiat@newton:~/hack/DDS$
>
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/online-bridge-hackathon/DDS/pull/65#pullrequestreview-445687352>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA3ZSBSQNIC7TO4KDIZ5KTR2XJWPANCNFSM4OSMHSMQ>
> .
>

suokko commented on Makefile:
```diff
@@ -18,6 +18,10 @@ GKE_ZONE ?= europe-west3-b
 
+LOCAL_URL := http://localhost:5000/api/dds-table/
+PROD_URL  := https://dds.prod.globalbridge.app/api/dds-table/
+URL       ?= ${PROD_URL}
```
I think default should be local testing. But it might not matter.

suokko commented on Makefile:
```diff
@@ -62,7 +66,9 @@ ensure_ns:
-curl_local:
+# We'd like to define a TEST_DEAL macro, but then the command-line output is
+# not as pretty.
+curl:
 	curl \
```
Do we need three targets?

I would think two would be enough.

tameware replied:
I agree that two would be better! It will depend on what strategy we agree on for constructing the targets. If we do end up with three I'll make the local target the default.

suokko commented on Makefile:
```diff
@@ -71,5 +77,11 @@ curl_local:
+	${URL}
+
+curl_local:
+	make curl URL=${LOCAL_URL}
```
Recursive make is wrong in most cases. It should be avoided pretty much always.

Better alternative would be something like
```
CURL_CMD := curl --header ...

curl_local:
    ${CURL_CMD} ${LOCAL_URL}

tameware replied:
I like your suggestion but it has a problem. When I use the recursive make I get a nicely formatted command line:

```
Ragnar:DDS-aw adamw$ make curl_prod
make curl URL=https://dds.prod.globalbridge.app/api/dds-table/
curl \
	--header "Content-Type: application/json" \
	--data \
	'{"hands": { \
		"S":["D3", "C6", "DT", "D8", "DJ", "D6", "CA", "C3", "S2", "C2", "C4", "S9", "S7"],    \
		"W":["DA", "S4", "HT", "C5", "D4", "D7", "S6", "S3", "DK", "CT", "D2", "SK", "H8"],    \
		"N":["C7", "H6", "H7", "H9", "CJ", "SA", "S8", "SQ", "D5", "S5", "HK", "C8", "HA"],    \
		"E":["H2", "H5", "CQ", "D9", "H4", "ST", "HQ", "SJ", "HJ", "DQ", "H3", "C9", "CK"] }}' \
	https://dds.prod.globalbridge.app/api/dds-table/
```

When I use either your suggestion or my earlier attempt with a DEAL macro, I get this:

```
Ragnar:DDS-aw adamw$ make curl_prod
curl --header "Content-Type: application/json" --data '{"hands": { "S":["D3", "C6", "DT", "D8", "DJ", "D6", "CA", "C3", "S2", "C2", "C4", "S9", "S7"], "W":["DA", "S4", "HT", "C5", "D4", "D7", "S6", "S3", "DK", "CT", "D2", "SK", "H8"], "N":["C7", "H6", "H7", "H9", "CJ", "SA", "S8", "SQ", "D5", "S5", "HK", "C8", "HA"], "E":["H2", "H5", "CQ", "D9", "H4", "ST", "HQ", "SJ", "HJ", "DQ", "H3", "C9", "CK"] }}'  https://dds.prod.globalbridge.app/api/dds-table/
```

Switching from **:=** to **=** did not make a difference.

Do you know a good way around this? 

I found Peter Miller's [Recursive Make Considered Harmful](http://www.real-linux.org.uk/recursivemake.pdf) article and I don't think his argument applies to a case like this.

suokko replied:
> I like your suggestion but it has a problem. When I use the recursive make I get a nicely formatted command line:
> 
> Switching from **:=** to **=** did not make a difference.
> 

You could split the command to multiple variables. Then use backslash after each to get multiline output.

> Do you know a good way around this?
> 
> I found Peter Miller's [Recursive Make Considered Harmful](http://www.real-linux.org.uk/recursivemake.pdf) article and I don't think his argument applies to a case like this.

I don't remember what was there. Example for relevant issues is that cmake Makefile backend can end up running same commands multiple times because it use very similar approach of recursing to same makefile multiple times. (One reason why I only started liking CMake after they added Ninja backend)

tameware replied:
Something like this?

```
curl_local:
	${CURL_CMD} \
	'{"hands": {     \
		${SOUTH},    \
		${WEST},     \
		${NORTH},    \
		${EAST}  }}' \
	${LOCAL_URL}
```

It works, but the recursive make call is shorter and seems cleaner.

tameware replied:
(These targets will not interact with CMake.)

tameware replied:
I've asked for suggestions:

https://stackoverflow.com/questions/62827211/how-to-embed-newlines-in-a-macro-inside-a-recipe

suokko replied:
> (These targets will not interact with CMake.)

True. But same risk are involved here as with recursion generated by CMake.

> I've asked for suggestions:
> 
> https://stackoverflow.com/questions/62827211/how-to-embed-newlines-in-a-macro-inside-a-recipe

I think the better solution to clean up makefile is the silent make the target idiom. http://make.mad-scientist.net/managing-recipe-echoing/

If makefile requires many calls to same tool then it could includes something like:
``
ifeq (V,)
SCURL = @echo CURL $@ && curl
else
SCURL = curl
endif

CURL_CMD = ${SCURL} --header ....
```
Does that default clean output with `make <targets> V=1` as a potential easy full command output for debugging?

suokko replied:
> I've asked for suggestions:
> 
> https://stackoverflow.com/questions/62827211/how-to-embed-newlines-in-a-macro-inside-a-recipe

Oops. My Makefile knowledge is fading. I forgot you can do something like
```
define CURL_CMD
curl --header ... \
 '{"hands": { ...\
... }'
$1
enddef

local_curl:
    $(call CURL_CMD,${URL})
```

tameware replied:
Neither of these gives the output I'm hoping for. _Make_ seems to turn newlines to spaces any time it expands a macro in a recipe.

The first StackExchange answer suggests:

_But to be honest, I would try to explain to your reviewer that fiddling with technically superfluous spaces and newlines in makefiles is lowering the quality of a makefile instead of raising it._

The recursive make solution is the only one where I've been able to get clean output, and I don't think the general objections apply since this command is simply "executable documentation" and does not involve any dependency analysis. Conceptually it could just as easily be accomplished with shell commands, but that would clutter our directory.

suokko replied:
> The recursive make solution is the only one where I've been able to get clean output, and I don't think the general objections apply since this command is simply "executable documentation" and does not involve any dependency analysis. Conceptually it could just as easily be accomplished with shell commands, but that would clutter our directory.

Makefiles either declare dependencies or become quickly hard to use. That is my experience with all Makefiles.

What do you say about following output?

```
CURL   http://localhost:5000/api/dds-table/
curl: (7) Failed to connect to localhost port 5000: Connection refused
make: *** [Makefile:88: curl_local] Error 7
```

It is produced from a silent makefile.

```
ifeq (V,1)
SCURL = curl
else
SCURL = @echo "CURL   ${1}" && curl
endif

define CURL_CMD
	${SCURL} \
	--header "Content-Type: application/json" \
	--data \
	'{"hands": { \
		"S":["D3", "C6", "DT", "D8", "DJ", "D6", "CA", "C3", "S2", "C2", "C4", "S9", "S7"], \
		"W":["DA", "S4", "HT", "C5", "D4", "D7", "S6", "S3", "DK", "CT", "D2", "SK", "H8"], \
		"N":["C7", "H6", "H7", "H9", "CJ", "SA", "S8", "SQ", "D5", "S5", "HK", "C8", "HA"], \
		"E":["H2", "H5", "CQ", "D9", "H4", "ST", "HQ", "SJ", "HJ", "DQ", "H3", "C9", "CK"] }}' \
	${1}
endef

curl_local:
	$(call CURL_CMD,${LOCAL_URL})

curl_prod:
	$(call CURL_CMD,${PROD_URL})
```

tameware replied:
This is the command output I'm hoping for. It shows the user exactly what _make_ is executing on his behalf.

```
Ragnar:DDS-aw adamw$ make curl_prod
make curl URL=https://dds.prod.globalbridge.app/api/dds-table/
curl \
	--header "Content-Type: application/json" \
	--data \
	'{"hands": { \
		"S":["D3", "C6", "DT", "D8", "DJ", "D6", "CA", "C3", "S2", "C2", "C4", "S9", "S7"],    \
		"W":["DA", "S4", "HT", "C5", "D4", "D7", "S6", "S3", "DK", "CT", "D2", "SK", "H8"],    \
		"N":["C7", "H6", "H7", "H9", "CJ", "SA", "S8", "SQ", "D5", "S5", "HK", "C8", "HA"],    \
		"E":["H2", "H5", "CQ", "D9", "H4", "ST", "HQ", "SJ", "HJ", "DQ", "H3", "C9", "CK"] }}' \
	https://dds.prod.globalbridge.app/api/dds-table/
```

How about I submit the recursive version and agree to change it if it ever becomes a problem?

tameware replied:
Actually, the first line of output is unnecessary. It's the curl command itself I want to show.

suokko replied:
Then there is non silent correctly formatted version without duplication. I forgot how to use make variables effectively.

```
curl_local_URL := http://localhost:5000/api/dds-table/
curl_prod_URL  := https://dds.prod.globalbridge.app/api/dds-table/
CURL_URL ?= ${$@_URL}

# We'd like to define a TEST_DEAL macro, but then the command-line output is
# not as pretty.
curl_local curl_prod:
	curl \
	--header "Content-Type: application/json" \
	--data \
	'{"hands": { \
		"S":["D3", "C6", "DT", "D8", "DJ", "D6", "CA", "C3", "S2", "C2", "C4", "S9", "S7"],    \
		"W":["DA", "S4", "HT", "C5", "D4", "D7", "S6", "S3", "DK", "CT", "D2", "SK", "H8"],    \
		"N":["C7", "H6", "H7", "H9", "CJ", "SA", "S8", "SQ", "D5", "S5", "HK", "C8", "HA"],    \
		"E":["H2", "H5", "CQ", "D9", "H4", "ST", "HQ", "SJ", "HJ", "DQ", "H3", "C9", "CK"] }}' \
	${CURL_URL}
```
`$@` gets expanded first to target and then the correct URL variable is expanded.

tameware replied:
That looks perfect! I'll try it and report back.

kiat-huang replied:
Interesting discussion, I'm at this step too. This PR works for me so far building and testing locally.

tameware replied:
Works nicely - thanks for persevering. I'll rebase later today.

kiat-huang replied:
This is working for me

CURL_ARGS ?= --header "Content-Type: application/json" \
    --data \
    '{"hands": { \
        "S":["D3", "C6", "DT", "D8", "DJ", "D6", "CA", "C3", "S2", "C2", "C4", "S9", "S7"],    \
        "W":["DA", "S4", "HT", "C5", "D4", "D7", "S6", "S3", "DK", "CT", "D2", "SK", "H8"],    \
        "N":["C7", "H6", "H7", "H9", "CJ", "SA", "S8", "SQ", "D5", "S5", "HK", "C8", "HA"],    \
        "E":["H2", "H5", "CQ", "D9", "H4", "ST", "HQ", "SJ", "HJ", "DQ", "H3", "C9", "CK"] }}'

curl:
    curl ${CURL_ARGS} ${URL}

suokko replied:
There is a minor difference:
```
$ make test1 test2 -f test.Makefile 
echo foo bar
echo foo \
	bar
foo bar
foo bar
$ cat test.Makefile 
ARGS ?= foo \
		bar

test1:
	echo ${ARGS}

test2:
	echo foo \
		bar
```
Adam wants to see the long command line split to multiple lines in the output. I agree with his wish. Figuring out the simplest solution was surprising hard.

kiat-huang commented:
> There's errors when I clone master, checkout this PR and run "make build"
> 
> ```
> kiat@newton:~/hack/DDS$ git checkout master
> Switched to branch 'master'
> Your branch is up-to-date with 'origin/master'.
> 
> kiat@newton:~/hack/DDS$ gh pr checkout 65
> Switched to branch 'new-prod-url'
> 
> kiat@newton:~/hack/DDS$ git st
> On branch new-prod-url
> nothing to commit, working tree clean
> 
> kiat@newton:~/hack/DDS$ git remote -v
> origin	git@github.com:online-bridge-hackathon/DDS.git (fetch)
> origin	git@github.com:online-bridge-hackathon/DDS.git (push)
> 
> kiat@newton:~/hack/DDS$ make build
> git submodule update --init --rebase
> fatal: repository '/home/kiat/Bridge/Hackathon/online-bridge-hackathon/libdds' does not exist
> fatal: clone of '/home/kiat/Bridge/Hackathon/online-bridge-hackathon/libdds' into submodule path '/home/kiat/Bridge/Hackathon/code/DDS/libdds' failed
> Failed to clone 'libdds'. Retry scheduled
> fatal: repository '/home/kiat/Bridge/Hackathon/online-bridge-hackathon/libdds' does not exist
> fatal: clone of '/home/kiat/Bridge/Hackathon/online-bridge-hackathon/libdds' into submodule path '/home/kiat/Bridge/Hackathon/code/DDS/libdds' failed
> Failed to clone 'libdds' a second time, aborting
> Makefile:29: recipe for target 'libdds-update' failed
> make: *** [libdds-update] Error 1
> 
> kiat@newton:~/hack/DDS$ lr
> total 68
> -rw-rw-r-- 1 kiat kiat     6 Jul  9 15:35 VERSION
> -rw-rw-r-- 1 kiat kiat   176 Jul  9 15:35 requirements.txt
> -rw-rw-r-- 1 kiat kiat 11415 Jul  9 15:35 LICENSE
> drwxrwxr-x 2 kiat kiat  4096 Jul  9 15:35 libdds
> -rw-rw-r-- 1 kiat kiat  1207 Jul  9 15:35 Dockerfile
> -rw-rw-r-- 1 kiat kiat   973 Jul  9 15:35 CONTRIBUTING.md
> -rw-rw-r-- 1 kiat kiat    25 Jul  9 15:35 _config.yml
> drwxrwxr-x 3 kiat kiat  4096 Jul  9 15:35 chart
> drwxrwxr-x 2 kiat kiat  4096 Jul  9 15:39 test
> drwxrwxr-x 2 kiat kiat  4096 Jul  9 15:39 src
> -rw-rw-r-- 1 kiat kiat  2552 Jul  9 15:39 README.md
> -rw-rw-r-- 1 kiat kiat  4561 Jul  9 15:39 README.gcp.md
> -rw-rw-r-- 1 kiat kiat  2459 Jul  9 15:39 Makefile
> drwxrwxr-x 2 kiat kiat  4096 Jul  9 15:39 frontend
> kiat@newton:~/hack/DDS$ 
> ```
This comment should be ignored - my environment for this branch was broken.

kiat-huang commented:
Added two commits for the Makefile.  Am done now.

suokko commented:
> Added two commits for the Makefile. Am done now.

I think these changes would deserve their own pull request after Adam's changes has been merged. It would be possible to have parallel pull request but I suspect there could be some conflicts.

kiat-huang commented:
> > Added two commits for the Makefile. Am done now.
> 
> I think these changes would deserve their own pull request after Adam's changes has been merged. It would be possible to have parallel pull request but I suspect there could be some conflicts.

Am happy with that. Now I'm working out how git can move the commits to a new PR.

tameware commented:
What's `dds-pr65`?

kiat-huang commented:
this fixes the Curl issue which you might want to include.

...it was in one of my commits which I've now moved out out of this PR with Pauli's help (thanks!)



```
CURL_ARGS ?= --insecure --header "Content-Type: application/json" \
    --data \
    '{"hands": { \
        "S":["D3", "C6", "DT", "D8", "DJ", "D6", "CA", "C3", "S2", "C2", "C4", "S9", "S7"],    \
        "W":["DA", "S4", "HT", "C5", "D4", "D7", "S6", "S3", "DK", "CT", "D2", "SK", "H8"],    \
        "N":["C7", "H6", "H7", "H9", "CJ", "SA", "S8", "SQ", "D5", "S5", "HK", "C8", "HA"],    \
        "E":["H2", "H5", "CQ", "D9", "H4", "ST", "HQ", "SJ", "HJ", "DQ", "H3", "C9", "CK"] }}'

curl:
    curl ${CURL_ARGS} ${URL} 
```

tameware commented:
Why —insecure? I thought we had addressed the server certificate issue.

On Fri, Jul 10, 2020 at 12:18 PM Kiat <notifications@github.com> wrote:

> this fixes the Curl issue which you might want to include.
>
> ...it was in one of my commits which I've now moved out out of this PR
> with Pauli's help (thanks!)
>
> CURL_ARGS ?= --insecure --header "Content-Type: application/json" \
>     --data \
>     '{"hands": { \
>         "S":["D3", "C6", "DT", "D8", "DJ", "D6", "CA", "C3", "S2", "C2", "C4", "S9", "S7"],    \
>         "W":["DA", "S4", "HT", "C5", "D4", "D7", "S6", "S3", "DK", "CT", "D2", "SK", "H8"],    \
>         "N":["C7", "H6", "H7", "H9", "CJ", "SA", "S8", "SQ", "D5", "S5", "HK", "C8", "HA"],    \
>         "E":["H2", "H5", "CQ", "D9", "H4", "ST", "HQ", "SJ", "HJ", "DQ", "H3", "C9", "CK"] }}'
>
> curl:
>     curl ${CURL_ARGS} ${URL}
>
> —
>
>
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/online-bridge-hackathon/DDS/pull/65#issuecomment-656815728>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABC4PYD6NMQIS3WXHO4773TR25LP3ANCNFSM4OSMHSMQ>
> .
>

kiat-huang commented:
On Prod the issue is fixed, not Dev.  And locally there is no SSL when testing.

For each of the 3 projects, DDS, Deal, Converter I've put PRs in with a Makefile, Docker and Chart files that deploy to Prod.  That was before I saw the curl targets and will merge them in.

I'm thinking maybe it is better to sort out these Prod PRs first before this one as they all have common structure (before they are all different) and they are a better basis to develop on, maybe marginally.

> Why —insecure? I thought we had addressed the server certificate issue.
> […](#)
> On Fri, Jul 10, 2020 at 12:18 PM Kiat ***@***.***> wrote: this fixes the Curl issue which you might want to include. ...it was in one of my commits which I've now moved out out of this PR with Pauli's help (thanks!) CURL_ARGS ?= --insecure --header "Content-Type: application/json" \ --data \ '{"hands": { \ "S":["D3", "C6", "DT", "D8", "DJ", "D6", "CA", "C3", "S2", "C2", "C4", "S9", "S7"], \ "W":["DA", "S4", "HT", "C5", "D4", "D7", "S6", "S3", "DK", "CT", "D2", "SK", "H8"], \ "N":["C7", "H6", "H7", "H9", "CJ", "SA", "S8", "SQ", "D5", "S5", "HK", "C8", "HA"], \ "E":["H2", "H5", "CQ", "D9", "H4", "ST", "HQ", "SJ", "HJ", "DQ", "H3", "C9", "CK"] }}' curl: curl ${CURL_ARGS} ${URL} — You are receiving this because you authored the thread. Reply to this email directly, view it on GitHub <[#65 (comment)](https://github.com/online-bridge-hackathon/DDS/pull/65#issuecomment-656815728)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABC4PYD6NMQIS3WXHO4773TR25LP3ANCNFSM4OSMHSMQ> .

tameware commented:
Local curls do not require the --insecure flag, at any rate neither on my Mac nor on the cloud shell. That said, I'm fine with leaving the flag set until we address the issue.

Your CURL_ARGS proposal doesn't get the job done. The command looks fine in the Makefile, but make runs the lines together when it echos the command before running it.

I'll merge this using @suokko's multiple target version. We can always update the URL constants later.

kiat-huang commented:
> Local curls do not require the --insecure flag, at any rate neither on my Mac nor on the cloud shell. That said, I'm fine with leaving the flag set until we address the issue.

Ack.
 
> Your CURL_ARGS proposal doesn't get the job done. The command looks fine in the Makefile, but make runs the lines together when it echos the command before running it.

Yes, I didn't understand that the Makefile gives you the formatting you want in this project, but not for the Deal project.
 
> I'll merge this using @suokko's multiple target version. We can always update the URL constants later.

suokko commented:
Looks good to me.

You remember to update the documentation too. Good!
